File: CNN-VAD-iOS/CNN_VAD/Audio IO/MovingAverageBuffer.m
//
//  MovingAverageBuffer.m
//  NC-2Ch
//
//  Created by Abhishek Sehgal on 5/12/17.
//  Copyright © 2017 SIPLab. All rights reserved.
//

#import "MovingAverageBuffer.h"

@interface MovingAverageBuffer()
@property (strong, nonatomic) NSMutableArray *queue;
@property (assign, nonatomic) NSUInteger period;
@property (assign, nonatomic) NSUInteger count;
@property (assign, nonatomic) float movingAverage;
@property (assign, nonatomic) float cumulativeAverage;
@end

@implementation MovingAverageBuffer

- (id)initWithPeriod:(NSUInteger)period {
    
    self = [self init];
    if(self){
        _period = period;
        _queue = [NSMutableArray array];
    }
    return self;
}

- (void)addDatum:(NSNumber *)datum {
    
    [self.queue insertObject:datum atIndex:0];
    
    float removed = 0;
    float datumf = [datum floatValue];
    
    if(self.queue.count > self.period) {
        removed = [[self.queue lastObject] floatValue];
        [self.queue removeLastObject];
    }
    
    self.movingAverage = self.movingAverage - (removed/self.period) + (datumf/self.period);
    
    self.cumulativeAverage = self.cumulativeAverage + (datumf - self.cumulativeAverage)/++self.count;
    
}

@end

File: CNN-VAD-Android/app/src/main/java/com/superpowered/frequencydomain/MovingAverageBuffer.java
package com.superpowered.frequencydomain;

import java.util.Queue;

/**
 * Created by abhisheksehgal on 10/6/17.
 */

public class MovingAverageBuffer {

    private float[] queue;
    private int period;
    public int count;
    public float movingAverage;
    public float cumulativeAverage;

    public MovingAverageBuffer() {
        period = 5;
        count = 0;
        movingAverage = 0;
        cumulativeAverage = 0;
        queue = new float[period];
    }

    public MovingAverageBuffer(int givenperiod){

        period = givenperiod;
        count  = 0;
        movingAverage = 0;
        cumulativeAverage = 0;
        queue = new float[period];

    }

    public void addDatum(float datum){
        float removed = queue[0];

        for (int i = 1; i < period; i++){
            queue[i-1] = queue[i];
        }

        queue[period-1] = datum;

        movingAverage = movingAverage - (removed/period) + (datum/period);
        cumulativeAverage = cumulativeAverage + (datum - cumulativeAverage)/++count;
    }
}
